### PR TITLE
Pure disarm/slow damage now decloaks units

### LINKED ISSUE #3932: Phandom does not decloak on disarm damage
Look at 15:40 on this cast: https://youtu.be/CyFENplD170?t=940

My Thunderbird nicked his Phantom and you can see the disarm, but it remained cloaked.

### PR DESCRIPTION
**Pull request by Arch-Shaman**
Fixes #3932:
* Pure disarm/slow damage now is considered damage by the decloak gadget.

### CODE REVIEW

**Comment by GoogleFrog** [comment]
Access to the WeaponDefs table is slow. The weapons that deal pure slow or disarm damage should be cached when the file is loaded.

**Comment by Arch-Shaman** [comment]
Understood. Will update it momentarily.

File: LuaRules/Gadgets/unit_decloak_damaged.lua [hunk 1]
```diff
@@ -320,6 +321,14 @@ function gadget:UnitCreated(unitID, unitDefID)
 end
 
 function gadget:Initialize()
+	local cp
+	for i=1, #WeaponDefs do -- cache pure slow and disarm damage weapons
+		cp = WeaponDefs[i].customParams or {}
```
Comment (sprunk) [inline LuaRules/Gadgets/unit_decloak_damaged.lua:326]: customparams is never nil
```suggestion
		cp = WeaponDefs[i].customParams
```

File: LuaRules/Gadgets/unit_decloak_damaged.lua [hunk 2]
```diff
@@ -320,6 +321,14 @@ function gadget:UnitCreated(unitID, unitDefID)
...
+		if cp.disarmdamageonly == '1' or cp.timeslow_onlyslow == '1' then
+			forceDecloakWeapons[i] = true
+		end
+	end
+	cp = nil
```
Comment (sprunk) [inline LuaRules/Gadgets/unit_decloak_damaged.lua:331]: not how the garbage collector works (`cp` will get garbage collected when it goes out of scope regardless, setting it to `nil` does not make this happen any sooner)
```suggestion
```

File: LuaRules/Gadgets/unit_decloak_damaged.lua [hunk 3]
```diff
@@ -320,6 +321,14 @@ function gadget:UnitCreated(unitID, unitDefID)
 end
 
 function gadget:Initialize()
+	local cp
```
Comment (sprunk) [inline LuaRules/Gadgets/unit_decloak_damaged.lua:324]: Should be local to the scope it's used in
```suggestion
```

**Comment by sprunk** [comment]
The most ideal solution would be to just filter out things which do zero damage (fake bomber trigger weapon etc) in the gadget handler and drop all the `damage > 0` checks everywhere but that is a larger and more dangerous task.

**Comment by sprunk** [comment]
This seems to be missing capture damage. Possibly there should be some `Spring.Utilities` function which returns weapons that do something despite zero nominal damage (see #4003).

**Comment by Arch-Shaman** [comment]
I'll open a new one with #4003's utilities.

**Pull request merged**